src/udev/udev-event.c

```diff
@@ -834,39 +833,28 @@ static int update_devnode(UdevEvent *event) {
 
         if (!uid_is_valid(event->uid)) {
                 r = device_get_devnode_uid(dev, &event->uid);
-                if (r == -ENOENT)
-                        event->uid = 0;
-                else if (r < 0)
+                if (r < 0 && r != -ENOENT)
                         return log_device_error_errno(dev, r, "Failed to get devnode UID: %m");
         }
 
         if (!gid_is_valid(event->gid)) {
                 r = device_get_devnode_gid(dev, &event->gid);
-                if (r == -ENOENT)
-                        event->gid = 0;
-                else if (r < 0)
+                if (r < 0 && r != -ENOENT)
                         return log_device_error_errno(dev, r, "Failed to get devnode GID: %m");
         }
 
         if (event->mode == MODE_INVALID) {
                 r = device_get_devnode_mode(dev, &event->mode);
                 if (r < 0 && r != -ENOENT)
                         return log_device_error_errno(dev, r, "Failed to get devnode mode: %m");
-                if (r == -ENOENT) {
-                        if (event->gid > 0)
-                                /* default 0660 if a group is assigned */
-                                event->mode = 0660;
-                        else
-                                /* default 0600 */
-                                event->mode = 0600;
-                }
         }
+        if (event->mode == MODE_INVALID && gid_is_valid(event->gid) && event->gid > 0)
+                /* If group is set, but mode is not set, "upgrade" mode for the group. */
+                event->mode = 0660;
```

Comment:
Why we should not "upgrade" mode when `gid == 0`?

Reply:
> Why we should not "upgrade" mode when gid == 0?

The kernel default is sufficient in this case (mode 0600). 

>  why my previous code introduces the issue.

Your code introduced the issue by adding `if (r == -ENOENT) uid=0;` and similar for gid and mode.

Reply:
> > Why we should not "upgrade" mode when gid == 0?
> 
> The kernel default is sufficient in this case (mode 0600).

What happens on 'change' uevent?

> > why my previous code introduces the issue.
> 
> Your code introduced the issue by adding `if (r == -ENOENT) uid=0;` and similar for gid and mode.

Yeah, I understand that point. Thanks.

Reply:
> > > Why we should not "upgrade" mode when gid == 0?
> > 
> > 
> > The kernel default is sufficient in this case (mode 0600).
> 
> What happens on 'change' uevent?

Ah, in that case, `r >= 0`, isn't it?